lib/ecto/type.ex

```diff
@@ -875,6 +884,94 @@ defmodule Ecto.Type do
     end
   end
 
+  @doc """
+  Checks if two terms are equal.
+
+  Depending on the given `type` performs a structural or semantical comparison.
+
+  ## Examples
+
+      iex> equal?(:integer, 1, 1)
+      true
+      iex> equal?(:decimal, Decimal.new("1"), Decimal.new("1.00"))
+      true
+
+  """
+  @spec equal?(t, term, term) :: boolean
+  def equal?(type, term1, term2) do
+    if requires_semantic_comparison?(type) do
+      do_equal?(type, term1, term2)
+    else
+      term1 == term2
+    end
+  end
+
+  defp requires_semantic_comparison?(:decimal),
+    do: true
+  defp requires_semantic_comparison?(:date),
+    do: false
+  defp requires_semantic_comparison?(type) when type in @calendar,
+    do: true
+  defp requires_semantic_comparison?(type) when type in @base,
+    do: false
+  defp requires_semantic_comparison?({:array, type}),
+    do: requires_semantic_comparison?(type)
+  defp requires_semantic_comparison?({:map, type}),
+    do: requires_semantic_comparison?(type)
+  defp requires_semantic_comparison?(module) when is_atom(module) do
+    function_exported?(module, :equal?, 2)
+  end
+
+  defp do_equal?(:decimal, %Decimal{} = term1, %Decimal{} = term2) do
+    Decimal.equal?(term1, term2)
+  end
+  defp do_equal?(type, %Time{} = term1, %Time{} = term2) when type in [:time, :time_usec] do
+    Time.compare(term1, term2) == :eq
+  end
+  defp do_equal?(type, %NaiveDateTime{} = term1, %NaiveDateTime{} = term2)
+       when type in [:naive_datetime, :naive_datetime_usec] do
+    NaiveDateTime.compare(term1, term2) == :eq
+  end
+  defp do_equal?(type, %DateTime{} = term1, %DateTime{} = term2)
+       when type in [:utc_datetime, :utc_datetime_usec] do
+    DateTime.compare(term1, term2) == :eq
+  end
+  defp do_equal?(type, term1, term2) when type in @base do
+    term1 == term2
+  end
+  defp do_equal?(module, term1, term2) when is_atom(module) do
+    module.equal?(term1, term2)
```

Comment:
Could `requires_semantic_comparison?/1` be removed if this part is changed to?

```elixir
defp do_equal?(module, term1, term2) when is_atom(module) do
  if function_exported?(module, :equal?, 2) do
    module.equal?(term1, term2)
  else
    term1 == term2
  end
end
```

In that case, `do_equal?` could be renamed to `equal?`

Reply:
No because this wont handle typed arrays and maps.

Reply:
Ah, this probably would work as you described; we are introducing the extra check to avoid semantical comparison unless we know it’s required as an optimization - vast majority of cases its not needed.

Reply:
Could we get similar performance benefits just be re-arranging the clauses?

```elixir
def equal?(type, term1, term2) when type in @base do
  term1 == term2
end
def equal?({:array, type}, xs, ys) do
  equal_list?(type, xs, ys)
end
def equal?({:map, _}, map1, map2) when map_size(map1) != map_size(map2) do
  false
end
def equal?({:map, type}, map1, map2) do
  equal_map?(type, Map.to_list(map1), Map.to_list(map2))
end
def equal?(type, %Time{} = term1, %Time{} = term2) when type in [:time, :time_usec] do
  Time.compare(term1, term2) == :eq
end
def equal?(type, %NaiveDateTime{} = term1, %NaiveDateTime{} = term2)
     when type in [:naive_datetime, :naive_datetime_usec] do
  NaiveDateTime.compare(term1, term2) == :eq
end
def equal?(type, %DateTime{} = term1, %DateTime{} = term2)
     when type in [:utc_datetime, :utc_datetime_usec] do
  DateTime.compare(term1, term2) == :eq
end
def equal?(:decimal, %Decimal{} = term1, %Decimal{} = term2) do
  Decimal.equal?(term1, term2)
end
def equal?(module, term1, term2) when is_atom(module) do
  if function_exported?(module, :equal?, 2) do
    module.equal?(term1, term2)
  else
    term1 == term2
  end
end
```

Reply:
The code above does unnecessary comparisons for eg `{:array, :integer}` but I see where youre going with this, I will investigate.

Reply:
@fertapric yes, that close (fyi: decimal and calendar types are in `@base` but that's a minor detail that's easy to fix)

We would still be inefficient for nested composites like `{:array, {:array, integer()}}` but to be fair that should be rather uncommon.

@josevalim do you have a preference?